tt_eager/tensor/types.hpp

```diff
@@ -309,19 +309,127 @@ struct BorrowedStorage {
 struct MultiDeviceHostStorage {
     std::vector<OwnedBuffer> buffers;
     std::vector<Shape> shapes;
-
+    std::mutex mtx;
     MultiDeviceHostStorage() = default;
+    MultiDeviceHostStorage(std::vector<OwnedBuffer> buffers_, std::vector<Shape> shapes_) : buffers(buffers_), shapes(shapes_) {}
+    MultiDeviceHostStorage(MultiDeviceHostStorage &&other) {
+        buffers = other.buffers;
+        shapes = other.shapes;
+    }
+
+    MultiDeviceHostStorage(const MultiDeviceHostStorage &other) {
+        buffers = other.buffers;
+        shapes = other.shapes;
+    }
+
+    MultiDeviceHostStorage &operator=(const MultiDeviceHostStorage &other) {
+        buffers = other.buffers;
+        shapes = other.shapes;
+        return *this;
+    }
+
+    MultiDeviceHostStorage &operator=( MultiDeviceHostStorage &&other) {
+        buffers = other.buffers;
+        shapes = other.shapes;
+        return *this;
+    }
+
+    bool operator == (const MultiDeviceHostStorage& other) {
+        return this->buffers == other.buffers and this->shapes == other.shapes;
+    }
+
     static constexpr auto attribute_names = std::make_tuple();
     const auto attribute_values() const { return std::make_tuple(); }
+
+    void insert_buffer_and_shape_for_device(Device* device, const OwnedBuffer buffer, const Shape shape) {
+        std::lock_guard<std::mutex> lock(mtx);
+        buffers[device->id()] = buffer;
+        shapes[device->id()] = shape;
+    }
+
+    OwnedBuffer get_buffer_for_device(Device* device) {
+        std::lock_guard<std::mutex> lock(mtx);
+        if (device->id() < buffers.size()) {
+            return buffers[device->id()];;
+        }
+        return OwnedBuffer();
+    }
+
+    Shape get_tensor_shape_for_device(Device* device) {
+        std::lock_guard<std::mutex> lock(mtx);
+        if (device->id() < shapes.size()) {
+            return shapes[device->id()];
+        }
+        return Shape({0xff, 0xff, 0xff, 0xff});
+    }
+
+    uint32_t num_buffers() {
+        std::lock_guard<std::mutex> lock(mtx);
+        return buffers.size();
+    }
 };
 
 struct MultiDeviceStorage {
-    std::vector<DeviceBuffer> buffers;
-    std::vector<Shape> shapes;
-
+    std::unordered_map<Device*, DeviceBuffer> buffers;
```

Comment:
The `MultiDeviceHostStorage` didn't need to change to use maps? Could we use the device-id as the key to the map, I think it'll be a lot easier to debug any issues down the road. Also, double-check you aren't iterating through these maps because no guarantee on ordering

Reply:
`MultiDeviceHostStorage` represents tensor storage that doesn't map to any devices, hence it didn't make sense to use a map here. We encode the device id as the index of the buffer inside the vector.

I took care of ordering for all cases where we construct host storage from device storage by inserting buffers at the required indices (derived from the device_id).